File: level_manager.py
from __future__ import annotations
import random
import pygame
from asteroid import Asteroid
from constants import (
    ASTEROID_MIN_RADIUS,
    ASTEROID_MAX_RADIUS,
    SCREEN_HEIGHT,
    SCREEN_WIDTH,
)

from typing import TYPE_CHECKING
if TYPE_CHECKING:
    from main import GameSession

class LevelManager:
    def __init__(self, session: GameSession) -> None:
        self.session = session
        self.current_level = 1
        self.state = "playing"  # "playing", "stage_clear"
        self.asteroids_remaining_to_spawn = 0
        self.speed_multiplier = 1.0
        self.wave_size = 3
        self.total_asteroids = 3
        self.bonus_credits = 0
        self.stage_clear_bonus = 0
        self.lives_bonus = 0

        # Define screen edges for spawning
        self.edges = [
            (
                pygame.Vector2(1, 0),
                lambda y: pygame.Vector2(-ASTEROID_MAX_RADIUS, y * SCREEN_HEIGHT),
            ),
            (
                pygame.Vector2(-1, 0),
                lambda y: pygame.Vector2(SCREEN_WIDTH + ASTEROID_MAX_RADIUS, y * SCREEN_HEIGHT),
            ),
            (
                pygame.Vector2(0, 1),
                lambda x: pygame.Vector2(x * SCREEN_WIDTH, -ASTEROID_MAX_RADIUS),
            ),
            (
                pygame.Vector2(0, -1),
                lambda x: pygame.Vector2(x * SCREEN_WIDTH, SCREEN_HEIGHT + ASTEROID_MAX_RADIUS),
            ),
        ]

        # Initialize Level 1
        self.start_level(1)

    def start_level(self, level: int) -> None:
        self.current_level = level
        self.session.current_level = level
        self.state = "playing"

        # Clean up any leftover entities from previous stage
        self.clean_up()

        # Procedural difficulty scaling calculations
        self.total_asteroids = 3 + (level - 1) * 2
        self.asteroids_remaining_to_spawn = self.total_asteroids

        # Wave size starts at 3, scaling by 1 every 3 levels, capped at 6
        self.wave_size = min(6, 3 + (level - 1) // 3)

        # Baseline speed starts at 1.0 and increases by 0.12 per level, capped at 2.5
        self.speed_multiplier = min(2.5, 1.0 + (level - 1) * 0.12)

        # Spawn the initial wave
        self.spawn_wave()

    def clean_up(self) -> None:
        # Safely remove all active asteroids, shots, and other transient objects
        # from the update and draw lists, leaving only the player and asteroid field.
        for sprite in list(self.session.updatable):
            if sprite is not self.session.player and sprite is not self.session.asteroid_field:
                sprite.kill()
        for sprite in list(self.session.drawable):
            if sprite is not self.session.player and sprite is not self.session.asteroid_field:
                sprite.kill()

    def spawn_wave(self) -> None:
        to_spawn = min(self.wave_size, self.asteroids_remaining_to_spawn)
        self.asteroids_remaining_to_spawn -= to_spawn

        # Select asteroid kinds (sizes/tiers) procedurally based on level
        if self.current_level == 1:
            kinds = [1, 2]
        elif self.current_level == 2:
            kinds = [1, 2, 3]
        elif self.current_level == 3:
            kinds = [2, 3]
        elif self.current_level == 4:
            kinds = [2, 3, 4]
        else:
            # Level 5 and beyond: expand up to tier 5 (increasing split potential/frequency)
            max_kind = min(5, 3 + (self.current_level - 5) // 2)
            kinds = list(range(max(1, max_kind - 2), max_kind + 1))

        for _ in range(to_spawn):
            edge = random.choice(self.edges)
            speed = random.randint(40, 100) * self.speed_multiplier
            velocity = edge[0] * speed
            velocity = velocity.rotate(random.randint(-30, 30))
            position = edge[1](random.uniform(0, 1))

            kind = random.choice(kinds)
            # Spawning adds the Asteroid automatically to containers set in create_session
            asteroid = Asteroid(position.x, position.y, ASTEROID_MIN_RADIUS * kind)
            asteroid.velocity = velocity

    def update(self, dt: float) -> None:
        if self.state == "playing":
            # Check if all active asteroids are cleared
            if len(self.session.asteroids) == 0:
                if self.asteroids_remaining_to_spawn > 0:
                    self.spawn_wave()
                else:
                    self.trigger_stage_clear()

    def trigger_stage_clear(self) -> None:
        self.state = "stage_clear"
        self.stage_clear_bonus = self.current_level * 100
        self.lives_bonus = self.session.lives * 50
        self.bonus_credits = self.stage_clear_bonus + self.lives_bonus
        self.session.credits += self.bonus_credits

    def advance_level(self) -> None:
        self.start_level(self.current_level + 1)

File: lib/game/level_manager.dart
import 'dart:math' as math;
import 'package:flame/components.dart';
import 'package:asteroids/constants.dart';
import 'package:asteroids/game/asteroids_game.dart';
import 'components/asteroid.dart';
import 'components/bullet.dart';
import 'components/bomb.dart';

class LevelManager {
  final AsteroidsGame game;
  int currentLevel = 1;
  String state = "playing"; // "playing", "stage_clear"
  
  int asteroidsRemainingToSpawn = 0;
  double speedMultiplier = 1.0;
  int waveSize = 3;
  int totalAsteroids = 3;
  int bonusCredits = 0;
  int stageClearBonus = 0;
  int livesBonus = 0;

  LevelManager(this.game) {
    startLevel(1);
  }

  void startLevel(int level) {
    currentLevel = level;
    game.currentLevel = level;
    state = "playing";

    // Clean up active asteroids, bullets, bombs, and explosions
    cleanUp();

    // Procedural difficulty scaling
    totalAsteroids = 3 + (level - 1) * 2;
    asteroidsRemainingToSpawn = totalAsteroids;

    // Wave size starts at 3, scaling by 1 every 3 levels, capped at 6
    waveSize = math.min(6, 3 + (level - 1) ~/ 3);

    // Baseline speed starts at 1.0 and increases by 0.12 per level, capped at 2.5
    speedMultiplier = math.min(2.5, 1.0 + (level - 1) * 0.12);

    // Spawn the initial wave
    spawnWave();
  }

  void cleanUp() {
    final asteroids = game.world.children.whereType<AsteroidComponent>();
    for (final asteroid in List<AsteroidComponent>.from(asteroids)) {
      asteroid.removeFromParent();
    }
    
    final bullets = game.world.children.whereType<BulletComponent>();
    for (final bullet in List<BulletComponent>.from(bullets)) {
      bullet.removeFromParent();
    }

    final bombs = game.world.children.whereType<BombComponent>();
    for (final bomb in List<BombComponent>.from(bombs)) {
      bomb.removeFromParent();
    }

    final explosions = game.world.children.whereType<BombExplosionComponent>();
    for (final explosion in List<BombExplosionComponent>.from(explosions)) {
      explosion.removeFromParent();
    }
  }

  void spawnWave() {
    final random = math.Random();
    final toSpawn = math.min(waveSize, asteroidsRemainingToSpawn);
    asteroidsRemainingToSpawn -= toSpawn;

    // Select asteroid kinds based on level
    List<int> kinds;
    if (currentLevel == 1) {
      kinds = [1, 2];
    } else if (currentLevel == 2) {
      kinds = [1, 2, 3];
    } else if (currentLevel == 3) {
      kinds = [2, 3];
    } else if (currentLevel == 4) {
      kinds = [2, 3, 4];
    } else {
      final maxKind = math.min<int>(5, 3 + (currentLevel - 5) ~/ 2);
      kinds = List.generate(
        maxKind - math.max<int>(1, maxKind - 2) + 1,
        (i) => math.max<int>(1, maxKind - 2) + i,
      );
    }

    final screenWidth = game.size.x;
    final screenHeight = game.size.y;
    final maxRadius = GameConstants.asteroidMaxRadius;

    for (var i = 0; i < toSpawn; i++) {
      // Pick a random edge: 0 = Left, 1 = Right, 2 = Top, 3 = Bottom
      final edgeIndex = random.nextInt(4);
      Vector2 direction;
      Vector2 position;
      final t = random.nextDouble();

      switch (edgeIndex) {
        case 0:
          direction = Vector2(1, 0);
          position = Vector2(-maxRadius, t * screenHeight);
          break;
        case 1:
          direction = Vector2(-1, 0);
          position = Vector2(screenWidth + maxRadius, t * screenHeight);
          break;
        case 2:
          direction = Vector2(0, 1);
          position = Vector2(t * screenWidth, -maxRadius);
          break;
        case 3:
        default:
          direction = Vector2(0, -1);
          position = Vector2(t * screenWidth, screenHeight + maxRadius);
          break;
      }

      final speed = (40.0 + random.nextDouble() * 60.0) * speedMultiplier;
      final deviationAngle = (-30.0 + random.nextDouble() * 60.0) * math.pi / 180.0;
      final velocity = direction.clone()..rotate(deviationAngle);
      velocity.scale(speed);

      final kind = kinds[random.nextInt(kinds.length)];
      final radius = GameConstants.asteroidMinRadius * kind;

      final asteroid = AsteroidComponent(
        radius: radius,
        position: position,
        velocity: velocity,
      );
      game.world.add(asteroid);
    }
  }

  void update(double dt) {
    if (state == "playing") {
      final asteroidsCount = game.world.children.whereType<AsteroidComponent>().length;
      if (asteroidsCount == 0) {
        if (asteroidsRemainingToSpawn > 0) {
          spawnWave();
        } else {
          triggerStageClear();
        }
      }
    }
  }

  void triggerStageClear() {
    state = "stage_clear";
    stageClearBonus = currentLevel * 100;
    livesBonus = game.lives * 50;
    bonusCredits = stageClearBonus + livesBonus;
    game.credits += bonusCredits;
    
    // Stop playing loop and show clear screen
    game.changeState(GameState.stageClear);
  }

  void advanceLevel() {
    startLevel(currentLevel + 1);
    game.changeState(GameState.playing);
  }
}
